metals/src/main/scala/scala/meta/internal/metals/DocumentHighlightProvider.scala

```diff
@@ -0,0 +1,162 @@
+package scala.meta.internal.metals
+
+import org.eclipse.lsp4j.DocumentHighlight
+import org.eclipse.lsp4j.DocumentHighlightKind
+import org.eclipse.lsp4j.TextDocumentPositionParams
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.mtags.DefinitionAlternatives.GlobalSymbol
+import scala.meta.internal.mtags.Semanticdbs
+import scala.meta.internal.mtags.Symbol
+import scala.meta.internal.semanticdb.Scala.Descriptor
+import scala.meta.internal.semanticdb.Scala.Symbols
+import scala.meta.internal.semanticdb.SymbolInformation
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.semanticdb.TextDocument
+
+final class DocumentHighlightProvider(
+    definitionProvider: DefinitionProvider,
+    semanticdbs: Semanticdbs
+) {
+
+  def documentHighlight(
+      params: TextDocumentPositionParams
+  ): java.util.List[DocumentHighlight] = {
+    val source = params.getTextDocument.getUri.toAbsolutePath
+    val result = semanticdbs.textDocument(source)
+
+    val highlights = for {
+      doc <- result.documentIncludingStale.toList
+      positionOccurrence = definitionProvider.positionOccurrence(
+        source,
+        params,
+        doc,
+        includeLastCharacter = true
+      )
+      occ <- positionOccurrence.occurrence.toList
+      alternatives = findAllAlternatives(doc, occ)
+      curr <- doc.occurrences
+      if curr.symbol == occ.symbol || alternatives(curr.symbol)
+      range <- curr.range
+      revised <- positionOccurrence.distance.toRevised(range.toLSP)
+      kind = if (curr.role.isDefinition) DocumentHighlightKind.Write
+      else DocumentHighlightKind.Read
+    } yield new DocumentHighlight(revised, kind)
+    highlights.asJava
+  }
+
+  private def findAllAlternatives(
+      doc: TextDocument,
+      occ: SymbolOccurrence
+  ): Set[String] = {
+
+    val symbolInfo = doc.symbols.find(_.symbol == occ.symbol)
+    symbolInfo match {
+      case Some(inf) =>
+        if (inf.isClass) Set(inf.symbol.dropRight(1) + ".")
+        else if (inf.isObject) Set(inf.symbol.dropRight(1) + "#")
+        else if (inf.isParameter && (inf.symbol.contains("apply") || inf.symbol
+            .contains("copy"))) {
+          parameterAlternatives(inf)
+        } else if (inf.isMethod) {
+          methodAlternatives(inf)
+        } else {
+          Set.empty
+        }
+      case None => Set.empty
+    }
+  }
+
+  private def methodAlternatives(info: SymbolInformation): Set[String] = {
```

Comment:
Looks good, I agree it came out nicely to not build on top of reference provider 👍